internal/installercache/installercache.go

```diff
@@ -80,22 +111,87 @@ func (rl *Release) Cleanup(ctx context.Context) {
 		"cached", rl.cached,
 		"extract_duration", rl.extractDuration,
 	)
+	if err := os.Remove(rl.Path); err != nil && !os.IsNotExist(err) {
+		return err
+	}
+	return nil
 }
 
 // New constructs an installer cache with a given storage capacity
-func New(cacheDir string, storageCapacity int64, eventsHandler eventsapi.Handler, log logrus.FieldLogger) *Installers {
+func New(config Config, eventsHandler eventsapi.Handler, diskStatsHelper metrics.DiskStatsHelper, log logrus.FieldLogger) (*Installers, error) {
+	if config.MaxCapacity > 0 && config.EvictionThresholdPercent == 0 {
+		return nil, errors.New("config.InstallerCacheEvictionThreshold must not be zero")
+	}
+	if config.MaxCapacity > 0 && config.MaxReleaseSize == 0 {
+		return nil, fmt.Errorf("config.MaxReleaseSize (%d bytes) must not be zero", config.MaxReleaseSize)
+	}
+	if config.MaxCapacity > 0 && config.MaxReleaseSize > config.MaxCapacity {
+		return nil, fmt.Errorf("config.MaxReleaseSize (%d bytes) must not be greater than config.MaxCapacity (%d bytes)", config.MaxReleaseSize, config.MaxCapacity)
+	}
 	return &Installers{
 		log:             log,
-		storageCapacity: storageCapacity,
-		cacheDir:        cacheDir,
 		eventsHandler:   eventsHandler,
-	}
+		diskStatsHelper: diskStatsHelper,
+		config:          config,
+	}, nil
 }
 
 // Get returns the path to an openshift-baremetal-install binary extracted from
 // the referenced release image. Tries the mirror release image first if it's set. It is safe for concurrent use. A cache of
 // binaries is maintained to reduce re-downloading of the same release.
 func (i *Installers) Get(ctx context.Context, releaseID, releaseIDMirror, pullSecret string, ocRelease oc.Release, ocpVersion string, clusterID strfmt.UUID) (*Release, error) {
+	for {
+		select {
+		case <-ctx.Done():
+			return nil, ctx.Err()
+		default:
+			release, err := i.get(releaseID, releaseIDMirror, pullSecret, ocRelease, ocpVersion, clusterID)
+			if err == nil {
+				return release, nil
+			}
+			_, isCapacityError := err.(*errorInsufficientCacheCapacity)
+			if !isCapacityError {
+				return nil, errors.Wrapf(err, "failed to get installer path for release %s", releaseID)
+			}
+			time.Sleep(i.config.ReleaseFetchRetryInterval)
+		}
+	}
+}
+
+func (i *Installers) getDiskUsageIncludingHardlinks() (uint64, error) {
+	usedBytes, _, err := i.diskStatsHelper.GetDiskUsage(i.config.CacheDir)
+	if err != nil && !os.IsNotExist(err) {
+		return 0, errors.Wrapf(err, "could not determine disk usage information for cache dir %s", i.config.CacheDir)
+	}
+	return usedBytes, nil
+}
+
+func (i *Installers) extractReleaseIfNeeded(path, releaseID, releaseIDMirror, pullSecret, ocpVersion string, ocRelease oc.Release, release *Release) error {
+	_, err := os.Stat(path)
+	if err == nil {
+		return nil // No error indicates that the release already exists
```

Comment:
This isn't really true because we return `nil` if we successfully fetch the release as well, right?

I think it's just as valid to just not have this comment.

Reply:
I kinda meant "no error from this call"
But I'll remove it

Reply:
Not any more, we made some changes.

Though I was simply referring to the meaning of err == nil on LN 171, perhaps the comment should have been there.